src/qh3/tls.py

```diff
@@ -1803,6 +1841,42 @@ def _client_handle_finished(self, input_buf: Buffer, output_buf: Buffer) -> None
         )
         next_enc_key = self.key_schedule.derive_secret(b"c ap traffic")
 
+        if self._certificate_request is not None:
+            with push_message(self.key_schedule, output_buf):
+                push_certificate(
+                    output_buf,
+                    Certificate(
+                        request_context=self._certificate_request.request_context,
+                        certificates=[
+                            (cert.public_bytes(Encoding.DER), b"")
+                            for cert in [self.certificate] + self.certificate_chain
+                            if cert is not None
+                        ],
+                    ),
+                )
+
+            signature_algorithm = negotiate(
+                self._signature_algorithms,
+                self._certificate_request.signature_algorithms,
+                AlertHandshakeFailure("No supported signature algorithm"),
+            )
+
+            # send certificate verify
+            signature = self.certificate_private_key.sign(
```

Comment:
Strictly speaking, one could arrive here having `self.certificate_private_key = None` unfortunately.
In that case, it is not optional. We should do something here.

Reply:
Looking again at the RFC, I believe the client should **only** respond to the `CertificateRequest` with no certificate, so I wrapped this piece of code, making sure this happens only if a certificate was passed

Reply:
It could be a misunderstanding, but for what you said I agreed.
It is not about that, 

A user may face an exception "None has no..." if `self.certificate_private_key` is actually None. e.g. passing keyfile = None or not setting it. So we have to find a way to not raise any exception in that case.

Reply:
This private key is paired to the certifictae provided, so if the user is expected to either supply them both or none. There is no point at proceeding to the next piece of code if not certificate was provided - but we should still validate the private key was supplied also

Reply:
exactly